arbstate/espresso_streamer.go

```diff
@@ -0,0 +1,355 @@
+	// the delayed message count is not finalized yet
+	if msg.DelayedMessagesRead < delayedMessageCount.Uint64() {
+		return false, fmt.Errorf("delayed message count is less than the finalized delayed message read count")
+	}
+
+	return true, nil
+}
+
+/* Verify the attestation quote */
+func (s *EspressoStreamer) verifyAttestationQuote(ctx context.Context, attestation []byte, userDataHash [32]byte) error {
+
+	method, ok := s.espressoTEEVerifierABI.Methods["verify"]
+	if !ok {
+		return fmt.Errorf("verify method not found")
+	}
+
+	calldata, err := method.Inputs.Pack(
+		attestation,
+		userDataHash,
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to pack calldata: %w", err)
+	}
+	fullCalldata := append([]byte{}, method.ID...)
+	fullCalldata = append(fullCalldata, calldata...)
+
+	_, err = s.l1Reader.Client().CallContract(
+		ctx,
+		ethereum.CallMsg{
+			To:   &s.espressoTEEVerifierAddr,
+			Data: fullCalldata,
+		}, nil,
+	)
+	if err != nil {
+		return fmt.Errorf("call to the contract failed: %w", err)
+	}
+
+	return nil
+}
+
+/**
+* Create a queue of messages from the hotshot to be processed by the node
+* It will sort the messages by the message index
+* and store the messages in `messagesWithMetadata` queue
+ */
+func (s *EspressoStreamer) queueMessagesFromHotshot(ctx context.Context) error {
+	if s.nextHotshotBlockNum == 0 {
+		// We dont need to check majority here  because when we eventually go
+		// to fetch a block at a certain height,
+		// we will check that a quorum of nodes agree on the block at that height,
+		// which wouldn't be possible if we were somehow given a height
+		// that wasn't finalized at all
+		latestBlock, err := s.espressoClients[0].FetchLatestBlockHeight(ctx)
+		if err != nil {
+			log.Warn("unable to fetch latest hotshot block", "err", err)
+			return err
+		}
+		log.Info("Started node at the latest hotshot block", "block number", latestBlock)
+		s.nextHotshotBlockNum = latestBlock
+	}
+
+	nextHotshotBlockNum := s.nextHotshotBlockNum
+
+	// TODO: should support multiple clients
+	arbTxns, err := s.espressoClients[0].FetchTransactionsInBlock(ctx, nextHotshotBlockNum, s.namespace)
+	if err != nil {
+		log.Warn("failed to fetch the transactions", "err", err)
+		return err
+	}
+	if len(arbTxns.Transactions) == 0 {
+		log.Info("No transactions found in the hotshot block", "block number", nextHotshotBlockNum)
+		return nil
+	}
+
+	for _, tx := range arbTxns.Transactions {
+		// Parse hotshot payload
+		attestation, userDataHash, indices, messages, err := arbutil.ParseHotShotPayload(tx)
+		if err != nil {
+			log.Warn("failed to parse hotshot payload, will retry", "err", err)
+			return err
+		}
+		// if attestation verification fails, we should skip this message
+		// Parse the messages
+		if len(userDataHash) != 32 {
+			log.Warn("user data hash is not 32 bytes")
+			continue
+		}
+		userDataHashArr := [32]byte(userDataHash)
+		err = s.verifyAttestationQuote(ctx, attestation, userDataHashArr)
```

Comment:
I am thinking we might need to specify the L1 block height for verifying attestation quotes. I am not sure how `pccs-onchain` works. But should that be better?